src/main/java/org/usfirst/frc4048/common/swervev3/io/SwerveModule.java

```diff
@@ -0,0 +1,109 @@
+    private final LoggableSystem<SwerveSteerMotorIO, SwerveSteerMotorInput> steerSystem;
+    private final LoggableSystem<SwerveAbsIO, SwerveAbsInput> absSystem;
+    private final PIDController drivePIDController;
+    private final ProfiledPIDController turningPIDController;
+    private final SimpleMotorFeedforward driveFeedforward;
+    private final SimpleMotorFeedforward turnFeedforward;
+    private double steerOffset;
+
+    public SwerveModule(SwerveDriveMotorIO driveMotorIO, SwerveSteerMotorIO steerMotorIO, SwerveAbsIO absIO, PID drivePid, PID turnPid, Gain driveGain, Gain turnGain, TrapezoidProfile.Constraints goalConstraint, String loggingKey) {
+        this.driveSystem = new LoggableSystem<>(driveMotorIO, new SwerveDriveMotorInput(), loggingKey);
+        this.steerSystem = new LoggableSystem<>(steerMotorIO, new SwerveSteerMotorInput(), loggingKey);
+        this.absSystem = new LoggableSystem<>(absIO, new SwerveAbsInput(), loggingKey);
+        drivePIDController = new PIDController(drivePid.getP(), drivePid.getI(), drivePid.getD());
+        turningPIDController = new ProfiledPIDController(turnPid.getP(), turnPid.getI(), turnPid.getD(), goalConstraint);
+        driveFeedforward = new SimpleMotorFeedforward(driveGain.getS(), driveGain.getV());
+        turnFeedforward = new SimpleMotorFeedforward(turnGain.getS(), turnGain.getV());
+        turningPIDController.enableContinuousInput(0, Math.PI * 2);
+    }
+
+    public SwerveModule(SwerveDriveMotorIO driveMotorIO, SwerveSteerMotorIO steerMotorIO, SwerveAbsIO absIO, SwervePidConfig pidConfig, String loggingKey) {
+        this(driveMotorIO, steerMotorIO, absIO, pidConfig.getDrivePid(), pidConfig.getSteerPid(), pidConfig.getDriveGain(), pidConfig.getSteerGain(), pidConfig.getGoalConstraint(), loggingKey);
+    }
+
+    public void setState(SwerveModuleState desiredState) {
+        double steerEncoderPosition = getSteerPosition();
+        SwerveModuleState state = SwerveModuleState.optimize(desiredState, new Rotation2d(steerEncoderPosition));
+        double driveSpeed = drivePIDController.calculate(driveSystem.getInputs().driveEncoderVelocity, (state.speedMetersPerSecond)) + driveFeedforward.calculate(state.speedMetersPerSecond);
+        double turnSpeed = turningPIDController.calculate(steerEncoderPosition, state.angle.getRadians()) + turnFeedforward.calculate(turningPIDController.getSetpoint().velocity);
+        driveSystem.getIO().setDriveVoltage(driveSpeed);
+        steerSystem.getIO().setSteerVoltage(turnSpeed * 12);
+    }
+
+    public SwerveModuleState getLatestState() {
+        return new SwerveModuleState(driveSystem.getInputs().driveEncoderVelocity, new Rotation2d(getSteerPosition()));
+    }
+
+    public void updateInputs() {
+        absSystem.updateInputs();
+        driveSystem.updateInputs();
+        steerSystem.updateInputs();
+        Logger.recordOutput(steerSystem.getKey() + "/steerOffset", steerOffset);
```

Comment:
Since steerOffset is an output, there is no need to log it (matches can also be replayed in simulation and can calculate this value. This is left over from when I was debugging.
Fixed